kernel: Add ROUND2_UP() and ROUND2_DOWN() macros

peter-mitsis opened the pull request:
Fixes #66454

Current Summary
==============
Re-worked this PR to just update __tdata_xx and __tbss_xx linker variables instead of trying to go the ROUND_UP/ROUND_DOWN() route (whose summary follows just for reference). Instead of rounding up __tdata_size and __tbss_size at runtime each time arch_tls_stack_setup() is called, the round up calculation has been moved into the linker script thread-local-storage.ld so that we do not need to perform the calculation at runtime at all.


Old Obsolete Summary
===================
Changed to ROUND2_UP and ROUND2_DOWN() (the ALIGN_xx names were already in use in sub-modules). 

Add new macros ROUND2_UP() and ROUND2_DOWN() that are similar to ROUND_UP() and ROUND_DOWN().

The significant difference between them is that unlike the ROUND_xx() macros, the ROUND2_xx() macros are restricted to alignments that are a power of two. (The ROUND_xx() macros used to be this way, but changed to allow arbitrary integer rounding mid 2023).

Whenever the alignment is both a power of two and a constant, then there ought to be no difference in the generated code when comparing ROUND2_xx() to ROUND_xx(). However, when the alignment is both a power of two and a variable, then the ROUND2_xx() macros can be expected to generate better performing code than the generic ROUND_xx() macros.

ceolin commented on include/zephyr/sys/util.h:
```diff
@@ -277,6 +277,22 @@ extern "C" {
+ *        which must be a power of 2.
+ */
+#define ALIGN_UP(x, align)                                   \
+	(((unsigned long)(x) + (unsigned long)(align) - 1) & \
```
shouldn't be better using `uintptr_t` ?

peter-mitsis replied:
> shouldn't be better using `uintptr_t` ?

You're probably right.

peter-mitsis commented:
Replaces the previously proposed ALIGN_xx() macros with ROUND2_UP() and ROUND2_DOWN() as ALIGN_UP() and ALIGN_DOWN() were already in use in other modules.

A type parameter has been added to the macro instead of forcing the math to cast values to an "unsigned long". This allows extra flexibility as the macro can now be used on 64-bit integer types on 32-bit architectures.

npitre commented:
Suggestion:

- Providing the type as an argument is nice, but this doesn't work for
  pointer types. Might be nicer to always cast to `uintptr_t` internally 
  and use the provided type for the final result.

_ As for allowing 64-bit values on 32-bit architectures, you could do:

```

#define ALIGN2_UP(type, val, align) \
    (sizeof(value) == 64 ? ALIGN2_64_UP(type, value, align) \
                         : ALIGN2_32_UP(type, value, align))
```

  This way, the internal cast could be `unsigned int` in the 32-bit case and
  `unsigned long long` in the 64-bit case.

- Pushing this further, the type could be substituted with
  `__typeof__(value)` without the need to provide it explicitly, making
  this macro even more useful and avoiding some mistakes:

```
#define ALIGN2_UP(val, align) \
    ((__typeof__(value)) \
     (sizeof(value) == 64 ? TYPED_ALIGN2_UP(unsigned long long, value, align) \
                          : TYPED_ALIGN2_UP(unsigned int, value, align)))
```

npitre commented:
Oh, and those `sizeof(value) == 64` above should be `sizeof(value) == 8`.

peter-mitsis commented:
@npitre - I really like your approach for keying off `sizeof(value)` -- very cool. I think I shall try that. :)

peter-mitsis commented:
@npitre - I encountered something that I was not expecting when trying out your proposal.

> error: invalid application of 'sizeof' to incomplete type 'char[]'

This stems from kernel_tls.h (see the 2nd commit) the variables of type `char[]` that we are trying to align. We could force a typecast, or use the proposed TYPED_xxx() macro, but ultimately I think that it just simplest and most straight-forward to put the onus of specifying the desired type onto the caller via ROUND2_xx(type, x, align).

npitre commented:
Well... here I have some tangential questions.

This code:

```
static inline size_t z_tls_data_size(void)
{
        size_t tdata_size = ROUND2_UP(size_t, __tdata_size, __tdata_align);
        size_t tbss_size = ROUND2_UP(size_t, __tbss_size, __tbss_align);

        return tdata_size + tbss_size;
}
```

Here `__tdata_size` is generated by the linker. It is even used within the
linker script as is to define `__tdata_end`. Why would the C code decide 
to create a size that is not representative of the actual binary? To me,
applying a round-up here looks just wrong. It should either be removed, or
a pretty good explanation in the preceding comment for its presence be added.
If an alignment is actually needed, it should be applied to the linker 
section directly in the linker script instead.

That being said ... here `__tdata_size` and `__tbss_align` are sort of an
abuse of the C language given that the linker may only create label adresses.
So, to communicate numerical value from the linker for consumption by C code,
we create fake arrays to which the linker assigns a fake address. This is 
therefore a rather exceptional and 
somewhat hackish case for which it would be wrong to tailor generic infrastructure for. In other words, `ROUND2_UP()`
should not bend backward to accommodate this case. Rather, you should have 
something like:

```
        size_t tdata_size = (uintptr_t)__tdata_size; /* from the linker */
```

This way, you convert up front the fake address into what it should be 
interpreted as i.e. an integer value. And the intermediate `uintptr_t` type
is necessary to avoid compiler warnings in case the width of `size_t` is
smaller than a pointer.

Whether or not you need to round it up afterwards is a different matter, 
but it is no longer a type issue.

andyross reviewed (commented):
So... just to be That Guy and repeat the point we all know but aren't discussing: there are **many** users of the existing ROUND_UP/DOWN() APIs, almost all of which predate the recent change to allow non-power-of-two and can thus be assumed to be called with power of two sizes.   Only a tiny handful of those are being updated to the "correct" API.

Basically, isn't this backwards?  Don't we want to structure this as (1) revert the change from last summer, (2) update/rename/whatever the existing power-of-two API, (3) add the non-power-of-two API for only the spots that actually require it (which/where/who?).

Otherwise all we're doing is making an unfortunately complicated situation even more complicated.  Frankly we never should have merged that first patch if it had runtime impact.

cfriedt commented:
Did I break something? 😬

peter-mitsis commented:
@andyross - You raised some good points. I've done a search through the git history and found that since that commit last summer, only a couple of the ROUND_UP/ROUND_DOWN additions appear to use non-power-of-two alignments. I am looking at renaming those instances to ROUND_STD_UP/DOWN, and reverting ROUND_UP/DOWN back to powers-of-two. Admittedly, I am not the most fond of the ROUND_STD_UP/DOWN naming, and am open to better ones.

@npitre - I've been playing more with you auto-size selection idea. Unfortunately, the macro expansion is observed to cause compiler warnings with pointers on 32-bit platforms due to a casting of 32-bit pointers to `unsigned long long` (mismatched sizes). Even though that branch gets optimized away by the compiler, the warning is still present.

npitre commented:
> @npitre - I've been playing more with you auto-size selection idea.
> Unfortunately, the macro expansion is observed to cause compiler warnings 
> with pointers on 32-bit platforms due to a casting of 32-bit pointers to 
> `unsigned long long` (mismatched sizes).

OK then. Here's the universal version:

```
#define ALIGN_DOWN(value, align) ({ \
        union { unsigned long long ll; __typeof__(value) x; } __v; \
        union { unsigned long long ll; __typeof__(value) x; } __a; \
        __v.x = (value); \
        __a.x = (__typeof__(value))((unsigned long long)(align) - 1); \
        __v.ll &= ~__a.ll; \
        __v.x; })
```

Works as well for both 32- and 64-bit types, on 32- and 64-bit architectures,
with pointers and integers, and either big or littl;e endian.

Here's the assembly I get for big endian ARM32 compilations:

```
void *foo(void *p)
{
        return ALIGN_DOWN(p, 16);
}

foo:
        bic     r0, r0, #15
        bx      lr
```

```
int bar(int x, int y)
{
        return ALIGN_DOWN(x, y);
}

bar:
        rsb     r1, r1, #0
        and     r0, r0, r1
        bx      lr
```

```
unsigned long long baz(unsigned long long x, int y)
{
        return ALIGN_DOWN(x, y);
}

baz:
        rsbs    r3, r2, #0
        asr     r2, r2, #31
        rsc     r2, r2, #0
        and     r0, r0, r2
        and     r1, r1, r3
        bx      lr
```

```
unsigned long long burp(unsigned long long x)
{
        return ALIGN_DOWN(x, 64);
}

burp:
        bic     r1, r1, #63
        bx      lr
```

You can't hardly get better than that.

npitre commented:
This being said, I still think it is wrong to perform any alignment at all 
in `z_tls_data_size()` per my earlier tangential comment.

peter-mitsis commented:
@npitre - Dang that is clever!

npitre commented:
> @npitre - Dang that is clever!

Not quite perfect yet though. This would be better:

```
#define ALIGN_DOWN(value, align) ({ \
        union { unsigned long long ll; __typeof__(value) x; } __v; \
        union { unsigned long long ll; __typeof__(value) x; } __a; \
        __v.x = (value); \
        __a.x = (__typeof__(value))((uintptr_t)(align) - 1); \
        __v.ll &= ~__a.ll; \
        __v.x; })
```

This one actually won't give warnings on pointers with variable alignment.
The only caveat is that you can't use an alignment bigger than what an 
`uintptr_t` can represent which should be quite reasonable.

For the "up" version, simply insert `__v.ll += ~__a.ll;` before the mask 
operation.

npitre commented:
I mean `__v.ll += __a.ll;` of course.  ;-)

henrikbrixandersen reviewed (commented):
We have other macros using "POW2" as part of their name to indicate that they work on power-of-two. Just a suggestion.

peter-mitsis commented:
@npitre - The union method is great for run time, but not so good for many legacy use cases where ROUND_UP/ROUND_DOWN used in variable definitions.

peter-mitsis committed:
linker: Round TLS sizes up in linker script

Instead of rounding up both __tdata_size and __tbss_size at runtime,
perform the calculation when the image is built.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis commented:
Re-vamped. This has been trimmed down to just what is needed to satisfy the original issue--no expansion of scope to include ROUND_UP()/ROUND_DOWN() changes.

The __tdata_size and __tbss_size are now rounded up to the next alignment when the image is built instead of (repeatedly) performing those calculations at runtime.

npitre commented:
You should cast `__tdata_siz` to `uintptr_t` first, then to `size_t`.
Otherwise you'll get warnings if pointer width is different from `size_t` width.

andyross reviewed (approved):
Oh, heh.  Are we not doing the rework to the rounding macros anymore?  No complaints about the linker script changes.

npitre commented:
Rounding macros can be reworked later if we still have a good reason for that.
